### PR TITLE
Tools: new benchmark for the hashkv performance

### LINKED ISSUE #8910: Analyze hash kv API performance
Create >1GB db file and measure how long it takes to get hash kvs.
This is critical for data corruption checking.

ref. https://github.com/coreos/etcd/pull/8554

### PR DESCRIPTION
**Pull request by spzala**
Provider a way to etcd users to check how long it takes to get
hashkv in user enviornment. The command output will provide time taken
to get hashkv along with db file size.

Fixed #8910

### CODE REVIEW

**Comment by spzala** [comment]
@gyuho hi, I tried it on a db file size around ~8GB and that's the output the new command provides, 
```
./bin/etcdctl check hashkv
HashKV : 4280873459
Endpoint : 127.0.0.1:2379
Time taken to get hashkv : 2.695825965s
DB size : 8269 MB
```
Hope it makes sense? The command will provide a way for any user to check the hashkv performance in their own environment.

**Comment by xiang90** [comment]
this is more a benchmark than a check. for a check, we should provide a criteria, and output failure or success.

**Comment by gyuho** [comment]
Agree with @xiang90. And we don't need a new command for hash kv, we already have:

https://github.com/coreos/etcd/blob/897c0559400ed6c7c004b0cf3fc3b56f55bfaa86/etcdctl/ctlv3/command/ep_command.go#L69-L74

For https://github.com/coreos/etcd/issues/8910, we just need to find out the cost of fetching hash kv.

**Comment by spzala** [comment]
@xiang90 I see, thanks much for your comments! @gyuho had a similar thoughts but I was wondering if it makes any sense with check!

**Comment by spzala** [comment]
@xiang90 @gyuho hello, does this approach of creating a benchmark cmd makes sense? Thanks!

**Comment by gyuho** [comment]
@spzala Yes! subcommand in benchmark is better. I will take a look at this PR this week. Thanks!

**Comment by spzala** [comment]
@gyuho sounds great. Thanks much!!

**Comment by spzala** [comment]
@gyuho initially I thought of moving some of write code into benchmark util method but wasn't sure if that could have make sense. Adding a sub-command sounds good, let me try it and will update the code. Thanks!

File: tools/benchmark/cmd/put.go [hunk 1]
```diff
@@ -150,3 +159,44 @@ func max(n1, n2 int64) int64 {
...
+	if !strings.HasPrefix(eps[0], `http://`) {
+		host = "http://" + eps[0]
+	}
+	resphash, errhash := clients[0].HashKV(ctx, host, epHashKVRev)
+	respstatus, errstatus := clients[0].Status(ctx, host)
+	cancel()
+	if errhash != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get the hashkv of endpoint %s (%v)\n", host, errhash)
+		panic(err)
+	} else {
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Do not need else?

File: tools/benchmark/cmd/put.go [hunk 2]
```diff
@@ -150,3 +159,44 @@ func max(n1, n2 int64) int64 {
...
+		results += fmt.Sprintf("  HashKV: %v \t\n", resphash.Hash)
+		results += fmt.Sprintf("  Endpoint: %v \t\n", host)
+		results += fmt.Sprintf("  Time taken to get hashkv: %v \t\n", time.Since(st).String())
+	}
+
+	// Get the db size
+	if errstatus != nil {
+		fmt.Println(fmt.Fprintf(os.Stderr, "Failed to get the status of endpoint %s (%v)\n", host, errstatus))
+		panic(err)
+	} else {
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Do not need else?

**Comment by gyuho** [comment]
@spzala Overall looks great. Could you fix CI and provide some sample outputs with GBs of data? Thanks a lot!

File: tools/benchmark/cmd/put.go (hunk 1 above)
Comment (spzala) [reply tools/benchmark/cmd/put.go]: yup, makes sense..not needed.

**Comment by spzala** [comment]
@gyuho thanks much, will be updating output example shortly.

**Comment by spzala** [comment]
@gyuho made a minor change in the latest commit in how output is being displayed (db size in GB vs MB I had in previous commit). Here are few related examples output, 
```
HaskKV Summary:
  HashKV: 1407003244
  Endpoint: http://127.0.0.1:2379
  Time taken to get hashkv: 917.035345ms
  DB size: 1.09GB
```
```
HaskKV Summary:
  HashKV: 1286658187
  Endpoint: http://127.0.0.1:2379
  Time taken to get hashkv: 1.609064019s
  DB size: 2.06GB
```
```
HaskKV Summary:
  HashKV: 4280873459
  Endpoint: http://127.0.0.1:2379
  Time taken to get hashkv: 4.971754313s
  DB size: 8.27GB
```

File: tools/benchmark/cmd/put.go [hunk 3]
```diff
@@ -150,3 +160,42 @@ func max(n1, n2 int64) int64 {
...
+	}
+	for i, ip := range eps {
+		eps[i] = strings.TrimSpace(ip)
+	}
+	st := time.Now()
+	var results string
+	ctx, cancel := context.WithCancel(context.Background())
+	clients[0].HashKV(ctx, eps[0], epHashKVRev)
+	if !strings.HasPrefix(eps[0], `http://`) {
+		host = "http://" + eps[0]
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Why do we need this?

Comment (spzala) [reply tools/benchmark/cmd/put.go]: @gyuho hi, I thought we check it by convention. It's not needed as you pointed out. I will remove it. Thanks!

File: tools/benchmark/cmd/put.go [hunk 4]
```diff
@@ -150,3 +160,40 @@ func max(n1, n2 int64) int64 {
...
+	results += fmt.Sprintf("  Time taken to get hashkv: %v \t\n", time.Since(st).String())
+
+	// Get the db size
+	if errstatus != nil {
+		fmt.Println(fmt.Fprintf(os.Stderr, "Failed to get the status of endpoint %s (%v)\n", host, errstatus))
+		panic(err)
+	}
+	dbsize := float64(respstatus.DbSize) / (1000 * 1000 * 1000)
+	results += fmt.Sprintf("  DB size: %vGB \t\n", strconv.FormatFloat(dbsize, 'f', 2, 64))
+	fmt.Println(results)
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: @spzala Let's use `humanize.Bytes` to string-ify db size. Then, this should be good to merge. Thanks!

**Comment by gyuho** [comment]
Can you also comment this result https://github.com/coreos/etcd/pull/9866#issuecomment-410565897 on 
https://github.com/coreos/etcd/issues/8910? Once this gets merged, we can close it.

File: tools/benchmark/cmd/put.go (hunk 4 above)
Comment (spzala) [reply tools/benchmark/cmd/put.go]: @gyuho thanks, this's good to know. Will make changes.

File: tools/benchmark/cmd/put.go [hunk 5]
```diff
@@ -150,3 +160,39 @@ func max(n1, n2 int64) int64 {
...
+	results += fmt.Sprintf("  HashKV: %v \t\n", resphash.Hash)
+	results += fmt.Sprintf("  Endpoint: %v \t\n", host)
+	results += fmt.Sprintf("  Time taken to get hashkv: %v \t\n", time.Since(st).String())
+
+	// Get the db size
+	if errstatus != nil {
+		fmt.Println(fmt.Fprintf(os.Stderr, "Failed to get the status of endpoint %s (%v)\n", host, errstatus))
+		panic(err)
+	}
+	results += fmt.Sprintf("  DB size: %v \t\n", humanize.Bytes(uint64(respstatus.DbSize)))
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Just `"  DB size: %s"`?

File: tools/benchmark/cmd/put.go [hunk 6]
```diff
@@ -150,3 +160,39 @@ func max(n1, n2 int64) int64 {
...
+		panic(err)
+	}
+	results += fmt.Sprintf("\nHaskKV Summary:\n")
+	results += fmt.Sprintf("  HashKV: %v \t\n", resphash.Hash)
+	results += fmt.Sprintf("  Endpoint: %v \t\n", host)
+	results += fmt.Sprintf("  Time taken to get hashkv: %v \t\n", time.Since(st).String())
+
+	// Get the db size
+	if errstatus != nil {
+		fmt.Println(fmt.Fprintf(os.Stderr, "Failed to get the status of endpoint %s (%v)\n", host, errstatus))
```
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Just do `fmt.Printf("...\n")`

Comment (spzala) [reply tools/benchmark/cmd/put.go]: Yup, thanks @gyuho!

File: tools/benchmark/cmd/put.go [hunk 7]
```diff
@@ -150,3 +160,39 @@ func max(n1, n2 int64) int64 {
...
+
+	resphash, errhash := clients[0].HashKV(ctx, host, epHashKVRev)
+	respstatus, errstatus := clients[0].Status(ctx, host)
+	cancel()
+	if errhash != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get the hashkv of endpoint %s (%v)\n", host, errhash)
+		panic(err)
+	}
+	results += fmt.Sprintf("\nHaskKV Summary:\n")
+	results += fmt.Sprintf("  HashKV: %v \t\n", resphash.Hash)
```
[the remark this answers is not in the document]
Comment (spzala) [reply tools/benchmark/cmd/put.go]: @gyuho my bad, yes we don't need tab. Thanks!!

**Pull request merged**